[1 - 2단계 방탈출 예약 대기] 코기(장재현) 미션 제출합니다.

jaehyeon2650 opened the pull request:
안녕하세요 찰리~! 코기입니다! 😊
이번 리뷰 잘 부탁드립니다!
이번 단계 미션의 변경 사항 커밋 목록 링크입니다 [링크](https://github.com/woowacourse/spring-roomescape-waiting/pull/186/files/5df47c1f355cb6efb6cee1c96b927539c95456ff..871b3781690ad1843d6305753cdc709891ebaea4)
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
[코기의 프롤로그](https://prolog.techcourse.co.kr/studylogs/4363)
## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
-  [x] 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.

 - [x] else 예약어를 쓰지 않는다.

- [ ]  모든 원시 값과 문자열을 포장한다.
-> 로직이 있는 경우만 포장하였습니다. 예) 맴버의 이름

 - [ ] 일급 컬렉션을 쓴다.

 - [x] 한 줄에 점을 하나만 찍는다.

 - [ ] 줄여 쓰지 않는다(축약 금지).

- [ ] 모든 엔티티를 작게 유지한다.

 - [ ] 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.

 - [ ] getter/setter/프로퍼티를 쓰지 않는다.
 -> dto <-> 도메인 객체 간 변환 과정에서 getter를 사용하였습니다.

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### Mockito
기존 jdbcTemplate을 사용하였을 때는 Fake 객체를 사용해서 Service 계층을 테스트하였습니다. 하지만 이번에 JPA로 교체하면서 기존에 사용하던 레포지토리 인터페이스를 사용하지 못해 만들어뒀던 Fake 객체를 사용하지 못하여 Mockito를 처음으로 사용해보았습니다. 현재 코드에서는 제대로 된 검증이 가능하지만(개발자가 모든 상황에 정확히 stub을 한 경우) 추후에 코드가 고쳐진다고 했을 때 이 테스트가 믿음직스럽지 않다고 느꼈습니다. 그래서 제가 낸 결론은 최대한 프로덕션과 비슷하게 사용하고 만약 비용적인 문제가 추가적인 문제로 통합테스트가 어려운 경우 Fake 객체를 사용해야한다고 생각이 들었습니다. 하지만 시간상의 이유로 Fake 객체도 못하겠다면 최후의 방법으로 Mockito를 사용할거라고 스스로 결정했습니다. 찰리의 생각은 어떤지 궁금합니다!

### 추후 리팩토링
예전 코드에서 제가 정의한 인터페이스를 제대로 활용을 못하여서 수정이 많았습니다. 예를 들어, 기존 서비스는 이전에 정의한 인터페이스를 의존하지만 JPA로 교체하면서 JpaMemberRepository 인터페이스를 의존하게 되어 서비스 관련 테스트들이 전부 깨졌습니다. 현재 코드에서는 해결하지 못했지만 페어와 다른 크루들과 토론을 하면서 합성 패턴을 사용하면 기존 Fake 테스트도 유지할 수 있다는 것을 깨달았습니다. 찰리의 피드백과 함께 기존 레포지토리 인터페이스를 활용하여 다시 리팩토링을 하려고 합니다.(현재는 기존 레포지토리 인터페이스, Fake 전부 삭제한 상태)

 
### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

jaehyeon2650 committed:
feat: setup project

jaehyeon2650 committed:
feat: 엔티티 매핑 및 레포지토리 구현

- 테마 레포지토리 jpa 레포지토리로 구현 및 테스트

jaehyeon2650 committed:
feat: 레포지토리 구현

- 멤버 레포지토리 jpa 레포지토리로 구현 및 테스트

jaehyeon2650 committed:
feat: 레포지토리 구현

- 예약시간(ReservationTime) 레포지토리 jpa 레포지토리로 구현 및 테스트

jaehyeon2650 committed:
feat: 레포지토리 구현

- 예약(Reservation) 레포지토리 jpa 레포지토리로 구현 및 테스트

jaehyeon2650 committed:
refactor: 테마 커스텀 레포지토리 생성

jaehyeon2650 committed:
refactor: 패키지 이동

jaehyeon2650 committed:
refactor: 서비스 의존성 교체

- jdbc -> jpa 레포지토리 교체

jaehyeon2650 committed:
remove: jdbc 관련 클래스 삭제

jaehyeon2650 committed:
test: 서비스 테스트 교체

- 통합 테스트 -> MOCK 테스트 변경

jaehyeon2650 committed:
docs: 내 예약 목록 조회 기능 명세 작성

jaehyeon2650 committed:
feat: 내 예약 조회 기능 추가

jaehyeon2650 committed:
test: 본인 예약 조회 테스트

jaehyeon2650 committed:
style: 코드 컨벤션 정리

jaehyeon2650 committed:
style: 코드 컨벤션 정리

jaehyeon2650 committed:
refactor: 기본 생성자 접근 제한자 변경

jaehyeon2650 committed:
docs: 구현한 기능 체크

Gomding commented on README.md:
```diff
@@ -18,6 +18,9 @@
         - 8일인 경우 1일부터 7일까지의 예약 건수 많은 순서대로 10개의 테마를 조회할 수 있다.
     - [x] `/` 요청 시 인기 테마 페이지를 응답한다.
         - `templates/index.html`
+- [x] 내 예약 목록을 조회하는 API 구현
+  - [x] 내 예약 목록 호출 시 GET /reservation-mine 요청한다.
+  - [x] `reservation-mine.html` 페이지가 응답된다.
```
추가된 요구사항도 잘 정리해주셨네요 👍 👍 

내 예약 목록 조회할 때 예외사항은 없었을까요? 😉

jaehyeon2650 replied:
현재는 url이 /reservation-mine이고, 로그인 된 토큰으로 요청을 보냅니다. 그렇기에 당장 생각나는 예외는 토큰을 보내지 않았거나, 유효하지 않은 토큰을 보냈을 정도라고 생각이 듭니다!
이는 ArugmentReolver를 통해서 인증이 진행되기에 이번 미션에서 요구사항이 아닌 이전 미션의 요구사항이라고 생각되어서 함께 적지 않았습니다! 혹시 이 외에 추가적인 예외 사항이 있을까요..?🤔 깊게 고민해봐도 잘 생각이 나지 않네요! 🥹

Gomding replied:
말씀하신 토큰에 대한 예외사항이 맞습니다!
로그인이 필요하다는것은 API의 요구사항은 아닐까요?

jaehyeon2650 replied:
아 맞습니다! 처음 이 readme를 보고 api 명세서 혹은 기능 사항을 보면 로그인을 안해도 사용할 수 있을거라고 착각할 수 있을거 같습니다! 추가하도록 하겠습니다👍

Gomding commented on src/main/java/roomescape/member/domain/Member.java:
사용하지 않는 메서드가 있군요! 😃 

```
    public Member assignId(Long id) {
        return new Member(id, name.getName(), email, password, role);
    }
```

jaehyeon2650 replied:
앗 그렇네요! 수정하도록 하겠습니다!

Gomding commented on README.md:
> 기존 jdbcTemplate을 사용하였을 때는 Fake 객체를 사용해서 Service 계층을 테스트하였습니다. 하지만 이번에 JPA로 교체하면서 기존에 사용하던 레포지토리 인터페이스를 사용하지 못해 만들어뒀던 Fake 객체를 사용하지 못하여 Mockito를 처음으로 사용해보았습니다. 현재 코드에서는 제대로 된 검증이 가능하지만(개발자가 모든 상황에 정확히 stub을 한 경우) 추후에 코드가 고쳐진다고 했을 때 이 테스트가 믿음직스럽지 않다고 느꼈습니다. 그래서 제가 낸 결론은 최대한 프로덕션과 비슷하게 사용하고 만약 비용적인 문제가 추가적인 문제로 통합테스트가 어려운 경우 Fake 객체를 사용해야한다고 생각이 들었습니다. 하지만 시간상의 이유로 Fake 객체도 못하겠다면 최후의 방법으로 Mockito를 사용할거라고 스스로 결정했습니다. 찰리의 생각은 어떤지 궁금합니다!

우선 Fake 객체를 사용하지 못한다고 하신 이유가 궁금하군요 :)

jaehyeon2650 replied:
기존 JDBCTemplate을 사용할 때 각 레포지토리 별로 인터페이스를 정의했습니다. 해당 인터페이스를 통해 inmemory 방식의 Fake 객체를 만들어서 Service를 단위테스트를 하였습니다.(현재는 삭제한 상태)  
JPA로 교체하면서 동적 쿼리 혹은 복잡한 쿼리가 사용될 경우에는 커스텀 레포지토리와 Spring Data Jpa 레포지토리를 상속받는 레포지토리를 만들었습니다. 
`public interface JpaReservationRepository extends JpaRepository<Reservation, Long>, ReservationCustomRepository {

    List<Reservation> findByDateAndTheme_Id(LocalDate date, Long themeId);

    boolean existsByTime_Id(Long timeId);

    boolean existsByDateAndTime_StartAtAndTheme_Id(LocalDate date, LocalTime time, Long themeId);

    boolean existsByTheme_Id(Long themeId);

    List<Reservation> findByMember_Id(Long id);
}`
해당 인터페이스를 구현하는 Fake 객체를 만드려면 JpaRepository에 명시되어있는 메서드를 모두 구현해야합니다. 실제로 해당 메서드를 사용하지 않는데도 구현을 해야해서 비효율적이라고 생각했습니다.
이를 PR 제출 몇시간 전까지 해결 못해서 Mockito를 사용했습니다. 하지만 다른 크루들과 얘기해보면서 어뎁터 패턴을 사용하면 해결할 수 있다해서 이를 이번 리팩토링을 통해서 해보려고합니다!

Gomding replied:
이렇게 구현하면 필요한 메서드만 사용하면서 Fake 메서드도 사용할 수 있을 것 같아요!

```java
public interface ReservationRepository {
// 필요한 인터페이스만 구현
}

public interface JpaReservationRepository extends ReservationRepository, JpaRepository<Reservation, Long> {

}

public class ReservationService {
    private final ReservationRepository reservationRepository;
}

public class FakeReservationRepository implements ReservationRepository {

}
```

+)
> 이를 PR 제출 몇시간 전까지 해결 못해서 Mockito를 사용했습니다.

그래도 테스트는 포기하지 않으셨군요 👍 👍 👍 👍

jaehyeon2650 replied:
해당 방법도 시도했습니다! 하지만 커스텀 레포지토리가 있는 경우 해당 방식을 사용하기 힘들다고 판단했습니다..ㅜ

Gomding commented on src/main/java/roomescape/member/domain/Member.java:
```diff
@@ -16,6 +33,10 @@ private Member(Long id, String name, String email, String password, Role role) {
 
+    public Member() {
+
+    }
```
JPA 를 위해 기본생성자를 선언해주셨군요
접근제어자를 public 으로 지정할 필요가 있을까요 :)

jaehyeon2650 replied:
헉! public만 가능한줄 알았습니다! 지금 보니 protected 생성자도 가능하네요! 현재 기본 생성자가 다른 패키지에서 사용하는 경우가 없기에 protected로 수정하면 좋을거같습니다! 생각치 못한 부분 알려주셔서 감사합니다 🥹

Gomding commented on src/main/java/roomescape/reservation/infrastructure/ReservationCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,64 @@
+import roomescape.reservation.domain.Reservation;
+
+@Repository
+public class ReservationCustomRepositoryImpl implements ReservationCustomRepository {
```
어떤 필요에 의해 구현하셨는지 궁금해요 :)

jaehyeon2650 replied:
@@#원래 기존에는 다음과 같이 쿼리 애노테이션을 사용해서 동적 쿼리를 처리하려고 하였습니다.

```
@Query("""
        SELECT r
        FROM Reservation r
        JOIN r.time t
        JOIN r.theme tm
        JOIN r.member m
        WHERE (:memberId IS NULL OR r.member.id = :memberId)
        AND (:themeId IS NULL OR r.theme.id = :themeId)
        AND (:dateFrom IS NULL OR r.date >= :dateFrom)
        AND (:dateTo IS NULL OR r.date <= :dateTo)
        ORDER BY r.id
        """)
    List<Reservation> findReservationsInConditions(
            @Param("memberId") Long memberId,
            @Param("themeId") Long themeId,
            @Param("dateFrom") LocalDate dateFrom,
            @Param("dateTo") LocalDate dateTo);
}
```

하지만 이 코드가 JpaRepository를 상속 받고 있는 인터페이스에 그대로 넣기에 긴 쿼리 문이 있고 만약 이런 동적 쿼리가 많아지고 해당 방식으로 구현한다면 인터페이스가 지저분해지고 가독성을 해칠거라고 생각했습니다.
처음 Spring Data JPA를 사용해봐서 다른 개발자가 직접 쿼리문을 작성할 수 있는 방법을 찾아보니 커스텀 레포지토리를 만드는 방법이 있었습니다. 해당 방식이 더 가독성과 유지보수에 유리할거라 생각해서 해당 방식을 선택하게되었습니다! 😊 제가 찾은 동적쿼리를 해결하는 방법은 이 2개만 찾았는데 더 좋은 방법 혹은 다른 방법이 있을까요? ㅎㅎ 궁금합니다!

Gomding replied:
다른 의도는 없었어요 ㅎㅎㅎ
정말 구현한 의도가 궁금했습니다 :)

Gomding commented on src/test/java/roomescape/theme/infrastructure/ThemeCustomRepositoryImplTest.java:
```diff
@@ -2,66 +2,51 @@
+        assertDoesNotThrow(() -> repository.deleteById(savedTheme.getId()));
     }
 
     @Test
```
잘못된 값으로 검증하고있는데 테스트는 성공하는군요🤔 

```
        List<String> names = reservations.stream()
                .map(Theme::getName)
                .toList();
        List<String> descriptions = reservations.stream()
                .map(Theme::getName)
                .toList();
        List<String> thumbnails = reservations.stream()
                .map(Theme::getName)
                .toList();
```

jaehyeon2650 replied:
헉 그렇네요! 운 좋게 통과가 된거군요.. 꼼꼼하게 살펴주셔서 감사합니다 🥹 수정하도록 하겠습니다.

Gomding commented on src/test/java/roomescape/theme/infrastructure/ThemeCustomRepositoryImplTest.java:
```diff
@@ -99,105 +84,98 @@ void find_all_test() {
-
-        Long themeId1 = repository.save(theme1);
-        Long themeId2 = repository.save(theme2);
+        em.persist(member);
+        em.persist(theme1);
+        em.persist(theme2);
+        em.persist(theme3);
```
EntityManager 를 활용해주셨군요 😃 
테스트에서 EntityManger를 사용하는데 장점이 많지만 단점은 없을까요?

jaehyeon2650 replied:
음.. 제가 생각한 단점으로는 현재는 given에서 필요한 데이터를 넣어주고 있습니다. 이렇게 엔티티 매니저를 사용하면 해당 레포지토리를 따로 사용안하더라도 편하게 등록할 수 있습니다.
만약 flush와 clear를 하지 않는다면 영속 상태인 객체에 대해 when 절에서 수정하거나 조회할 때 실제 데이터베이스에 쿼리를 날리지 않고 영속성 컨텍스트에 있는 객체를 업데이트하고 추후에 끝나면 쿼리를 모두 날리기에 실제 동작과는 다르게 진행된다고 생각했습니다.
이 외에는 단점이 생각나지 않네요..ㅜ 혹시 다른 단점이 있을까요?

Gomding replied:
키워드를 잘못 드렸네요 😢 
단점보다는 주의할점인데

말씀하신대로 flush 와 clear 를 제대로 활용하지 않으면 테스트가 오염될 수 있다는 것이에요!
해당 부분에 대한 지식도 필요하구요

Gomding commented on README.md:
> 예전 코드에서 제가 정의한 인터페이스를 제대로 활용을 못하여서 수정이 많았습니다. 예를 들어, 기존 서비스는 이전에 정의한 인터페이스를 의존하지만 JPA로 교체하면서 JpaMemberRepository 인터페이스를 의존하게 되어 서비스 관련 테스트들이 전부 깨졌습니다. 현재 코드에서는 해결하지 못했지만 페어와 다른 크루들과 토론을 하면서 합성 패턴을 사용하면 기존 Fake 테스트도 유지할 수 있다는 것을 깨달았습니다. 찰리의 피드백과 함께 기존 레포지토리 인터페이스를 활용하여 다시 리팩토링을 하려고 합니다.(현재는 기존 레포지토리 인터페이스, Fake 전부 삭제한 상태)


기존 인터페이스를 활용하지 못했던것은 시간이 없어서였을까요?
아직 방법을 찾지 못해서 였을까요?

jaehyeon2650 replied:
페어 프로그래밍을 하면서 페어와 고민해봤지만 그때 당시에는 해결방법은 찾지 못했습니다. 그래서 기존 JDBC 관련 코드를 지우고 Fake 객체를 이용한 테스트도 모두 직접 고쳤습니다.
PR 제출 시간 몇 시간 전에 다른 크루들과 토론하면서 어뎁터 패턴을 통해 해결할 수 있다는 것을 알 수 있었습니다. 아직 패턴에 대해 잘 몰라서 최소한 공부를 해보고 적용하고 싶었습니다. 하지만 시간상 바로 적용하지 못할거같아서 리팩토링 때 도전해보고자 합니다!

Gomding commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -8,13 +13,26 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
+    private Member member;
+
+    private LocalDate date;
+
+    @ManyToOne
+    private ReservationTime time;
+
+    @ManyToOne
+    private Theme theme;
```
`@ManyToOne` 으로 관계를 잘 매핑해주셨네요 👍 
현재 예약을 조회할 때 쿼리가 몇개 실행되는지 확인해보시겠어요? 😃

jaehyeon2650 replied:
ManyToOne은 기본으로 지연로딩을 지원하네요! 
` List<Reservation> findByMember_Id(Long id);`
그래서 쿼리 문을 조회하니 여러 개의 쿼리문이 나갔습니다. 현재 init.sql을 토대로보면 다음과 같았습니다.

> INSERT INTO reservation (member_id, date, time_id, theme_id) VALUES (1, '2025-04-28', 1, 1);
INSERT INTO reservation (member_id, date, time_id, theme_id) VALUES (1, '2025-04-28', 2, 1);

현재 만약 memberId가 1인 예약을 조회하게 되면 조건절에 있는 member는 조인해서 가져오고 나머지는 select 문으로 가져옵니다. 지금 예약 상태를 보면 theme는 아이디가 1인 예약이 하나만 있고 time은 아이디가 1,2인 두 경우가 있습니다. 그래서 결국 쿼리 문은 member와 reservation 조인문, theme select문 하나, time select문 2개로 총 4개가 나갑니다.
조금 더 실험을 해보기 위해 추가적으로 memberId가 1인 예약을 더 추가해보고 실험을 해봤습니다. 그러니 theme, time의 종류마다 select 쿼리가 나가는 것을 확인할 수 있었습니다. 이 문제를 조사해보니 Eager N+1 문제라고 하더군요! 그래서 해당 문제를 해결할 방법을 찾아보니 개발자가 jpql로 fetch 조인하기, `@EntityGraph `애노테이션을 통해 함께 조인을 해서 가져오는 방법이 있었습니다!
찰리 덕분에 몰랐던 개념에 대해서 알 수 있었고, 재밌게 학습했습니다! 감사합니다 😊 혹여나 제가 잘 못 이해한 개념이 있다면 알려주시면 감사할거같습니다!

Gomding replied:
잘 학습하신것 같네요 👍 👍 👍 

추가로 fetchType 에 대해서는 EAGER 를 유지하셨는데 이유가 궁금해요!
LAZY 와 어떤 차이가 있을까요? 😃

jaehyeon2650 replied:
Eager는 데이터베이스에서 조회하여 객체를 생성할 때 연관된 객체를 모두 끌어오는 설정입니다. Lazy는 그와 반대로 개발자가 연관된 객체를 직접적으로 사용할 때 그 떄 되어서야 쿼리를 날려서 가져옵니다!
현재 프로젝트만 보면 웬만한 로직에서 ReservationTime과 Theme를 함께 쓰고 있어서 Eager로 선택했습니다! 하지만 연관된 객체가 많아진다면 사용하지 않은데 쿼리를 날려서 성능을 낮출 수 있다고 생각합니다! 다음 미션을 하면서 상황을 보고 Lazy로 바꾸도록 하겠습니다! 😊

Gomding commented on src/main/java/roomescape/reservation/presentation/ReservationController.java:
```diff
@@ -63,4 +66,10 @@ public ResponseEntity<ExceptionResponse> noMatchDateType(final HttpServletReques
         );
         return ResponseEntity.badRequest().body(exceptionResponse);
     }
+    // TODO : URL
```
어떤 의미의 TODO 주석일까요? 😄

jaehyeon2650 replied:
/mine URL을 고치고 나서 TODO 주석을 까먹고 안지운거 같습니다..ㅜ 혼란만 야기했네요. 지우도록 하겠습니다!

Gomding commented on src/main/java/roomescape/reservation/infrastructure/JpaReservationRepository.java:
```diff
@@ -0,0 +1,20 @@
+
+public interface JpaReservationRepository extends JpaRepository<Reservation, Long>, ReservationCustomRepository {
+
+    List<Reservation> findByDateAndTheme_Id(LocalDate date, Long themeId);
```
언더바는 필요한 걸까요?

jaehyeon2650 replied:
Reservation 객체의 getter 메서드를 보면 내부 객체 ReservationTime, Theme의 getter 메서드는 만들지 않았습니다! 그 이유는 굳이 외부에서 Reservation 객체 구성 요소를 정확히 알아야하는지 의문이였기에 메시지를 통해 구체적인 반환 값을 get하는 메서드를 각각 만들었습니다.
하지만 언더바를 제외하고 jpaRepository 메서드의 이름을 짓게 되니 정확히 매핑이 되지 않는다는 오류가 발생했습니다. 그 이유를 조사하니 ThemeId라고 하면 getter가 있어야 Theme 객체의 id라고 정확히 매핑이 된다고 하더라고요! 만약 getter가 없는 경우 명시적으로 언더바를 사용해서 Theme 객체의 Id라고 명시해야 정확한 매핑이 된다고 합니다!
기존 getter 사용방식을 유지하고 싶어 언더바형태로 놔두게되었습니다!

Gomding replied:
메서드 이름이 잘 사용하지 않는 형태라 고민이 되네요 😃

Gomding commented on src/main/java/roomescape/reservation/dto/response/MyReservationResponse.java:
```diff
@@ -0,0 +1,21 @@
+                reservation.getId(),
+                reservation.getThemeName(),
+                reservation.getDate().toString(),
+                reservation.getReservationTime().toString(),
```
LocalTime 을 그대로 사용하지않고 문자열로 변경하는 이유가 궁금해요!

jaehyeon2650 replied:
앗.. 그렇네요! 솔직하게 말씀드리면 생각없이 응답이 JSON이라 String으로 한거같습니다. 생각해보니 LocalTime, LocalDate를 사용하면 자동으로 기본 형식을 맞춰서 반환을 해주네요! 수정하도록 하겠습니다.

Gomding reviewed (changes requested):
안녕하세요 코기!
방탈출 예약 대기 미션 함께하게된 찰리입니다~
미션 잘 구현해주셨네요 👍 👍 
몇가지 코멘트 남겼으니 확인부탁드려요!

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요 🙇

jaehyeon2650 reviewed (commented):
찰리의 리뷰 덕분에 JPA에 대해 추가적으로 즐겁게 학습할 수 있었습니다! 또한 꼼꼼한 리뷰 덕분에 놓치거나 실수한 부분을 바로 잡을 수 있었습니다! 감사합니다 😊

jaehyeon2650 committed:
style: TODO 주석 제거

jaehyeon2650 committed:
refactor: Jpa관련 레포지토리 어뎁터 패턴 적용

- 이전 인터페이스 복구(MemberRepository,ThemeRepository, ReservationRepository, ReservationTimeRepository)
- 이전 fake 테스트 복구
- 기존 Mockito 테스트 삭제

jaehyeon2650 committed:
refactor: ReservationRepository fetch join 적용

jaehyeon2650 committed:
refactor: MyReservationResponse 필드 타입 변경

- String -> LocalDate, LocalTime

jaehyeon2650 committed:
refactor: 코드 컨벤션 정리

Gomding commented on README.md:
저번에 얘기를 못했는데 테스트 개수가 엄청나군요..
쉽지 않았을 텐데 꼼꼼하게 작성하느라 고생하셨습니다!! 👍 👍 👍 💯 

![image](https://github.com/user-attachments/assets/11c69733-7afd-45f7-83d9-d1f702836bf7)

jaehyeon2650 replied:
감사합니다 😊 하지만 찰리 덕분에 테스트는 많지만 중간 실수로 통과한 테스트를 확인할 수 있었습니다!

Gomding commented on src/main/java/roomescape/theme/domain/Theme.java:
```diff
@@ -0,0 +1,92 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String name;
+
+    private String description;
+
+    private String thumbnail;
```
`@Column` 애너테이션을 활용해서 메타정보를 표시할 수 있곘군요 :)

jaehyeon2650 replied:
컬럼 에너테이션을 활용해보도록 하겠습니다!

Gomding commented on src/main/java/roomescape/theme/domain/Theme.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private static void validate(final String name, final String description, final String thumbnail) {
+        if (name == null || name.isBlank() || name.length() > MAX_NAME) {
+            throw new IllegalArgumentException("이름은 1글자 이상, 255글자 이하여야합니다.");
+        }
+        if (description == null || description.isBlank() || description.length() > MAX_DESCRIPTION) {
+            throw new IllegalArgumentException("설명은 1글자 이상, 255글자 이하여야합니다.");
+        }
+        if (thumbnail == null || thumbnail.isBlank() || thumbnail.length() > MAX_THUMBNAIL) {
+            throw new IllegalArgumentException("썸네일 URI는 1글자 이상, 255글자 이하여야합니다.");
+        }
```
현재 구현에서 DB 에는 255 글자 초과하여 들어갈 수 있는지 궁금해요!
또는 255 글자까지 허용하고 있을까요?

jaehyeon2650 replied:
헉! 현재 h2 데이터베이스르 사용하고 있는데 테스트를 해보니 한글, 영어 모두 글자가 255자해도 안되네요!
바이트 단위로 바꿔서 계산하기 때문에 실제 길이보다 더 길게 측정이 되네요..! 이를 고려해서 길이 검증을 바꿔야할거 같아요!
놓친 부분 다시 한번 찰리가 잡아주셨네요! 감사합니다 😊

Gomding commented on src/main/java/roomescape/theme/domain/Theme.java:
```diff
@@ -0,0 +1,92 @@
+    protected Theme() {
+    }
+
+    public static Theme createWithoutId(final String name, final String description, final String thumbnail) {
+        validate(name, description, thumbnail);
+        return new Theme(null, name, description, thumbnail);
+    }
+
+    public static Theme createWithId(final Long id, final String name, final String description,
+                                     final String thumbnail) {
+        validate(name, description, thumbnail);
+        return new Theme(Objects.requireNonNull(id), name, description, thumbnail);
+    }
```
validate 메서드를 생성자 내부에서 호출하면 객체 생성 시 공통으로 검사가 될 것 같은데
앞으로 추가될 팩터리 메서드도 자연스럽게 적용될것이구요
검사가 누락되는 실수도 방지할 수 있겠네요
코기는 어떻게 생각하시나요?

jaehyeon2650 replied:
원래 처음 의도는 팩터리 메서드 마다 다르게 검증 할 수도 있을거같아서 해당 선택을 하였는데 생각을 해보니 현재 검증은 데이터베이스와 관련된 검증으로 모두 같은 검증을 할거같네요!
찰리 말대로 생성자를 통해 검증해서 누락 방지를 하는게 좋아보입니다! 놓친 부분 확인 감사합니다 😊

Gomding commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    public ReservationResponse createReservation(final ReservationRequest request, final Long memberId) {
+        ReservationTime time = reservationTimeRepository.findById(request.timeId())
+                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 시간입니다."));
+        Theme theme = themeRepository.findById(request.themeId())
+                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 테마입니다."));
+
+        Optional<Member> findMember = memberRepository.findById(memberId);
+        if (findMember.isEmpty()) {
+            throw new IllegalArgumentException("존재 하지 않는 유저입니다.");
+        }
+
+        Reservation reservation = Reservation.createWithoutId(dateTime.now(), findMember.get(), request.date(), time,
+                theme);
+
+        if (reservationRepository.hasSameReservation(reservation)) {
+            throw new IllegalArgumentException("이미 예약이 존재합니다.");
+        }
```
검증 순서를 조정하면 
검증에서 예외 발생 시 DB 조회 비용을 아낄 수 있을 것 같네요 :)

jaehyeon2650 replied:
현재 hasSameReservation에서 Reservation 객체를 받고 있어서 결국 Reservation을 객체를 생성해야하는데 검증 순서만 바꿔서 조회 비용을 아끼는 방법이 떠오르지 않습니다..ㅠ
만약 순서만 바꾸는 것이 아니라 hasSameReservation의 파라미터를 Reservation 객체가 아니라 필드 값들을 받는다면 가능할거같습니다..! 혹시 어떻게 순서만 바꿔서 조회 비용을 아낄 수 있을까요..?

Gomding commented on src/main/java/roomescape/reservation/dto/response/MyReservationResponse.java:
```diff
@@ -0,0 +1,23 @@
+                reservation.themeName(),
+                reservation.getDate(),
+                reservation.reservationTime(),
+                "예약"
```
다음 단계에서는 문자열 하드코딩 대신 다른 방법으로 리팩터링 예정이겠죠? 😉

jaehyeon2650 replied:
네 맞습니다! 다음 미션에서 진행하기 위해 하드코딩된 상태로 남겨둔 것입니다!

Gomding reviewed (approved):
안녕하세요 코기!
이번 단계 잘 구현해주셔서 이만 머지하겠습니다 :)
추가로 남긴 코멘트는 다음 단계 구현하면서 함께 고민해주세요~

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요!!
다음 단계도 화이팅입니다 👏
